src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,388 @@
+        private static char propertyDelimiter = ':';
+
+        internal static  DefaultSkipTokenHandler Instance = new DefaultSkipTokenHandler();
+
+        /// <summary>
+        /// Constructor for DefaultSkipTokenHandler - Sets the Property Delimiter
+        /// </summary>
+        public DefaultSkipTokenHandler()
+        {
+            IsDeltaFeedSupported = false;
+        }
+
+        /// <summary>
+        /// Constructor for Unit testing purposes - Sets the Property Delimiter
+        /// </summary>
+        public DefaultSkipTokenHandler(char delimiter)
+            : this()
+        {
+            propertyDelimiter = delimiter;
+        }
+
+        /// <summary>
+        /// Returns the URI for NextPageLink
+        /// </summary>
+        /// <param name="baseUri">BaseUri for nextlink. It should be request URI for top level resource and navigationlink for nested resource.</param>
+        /// <param name="instance">Instance based on which SkipToken value will be generated.</param>
+        /// <param name="pageSize">Maximum number of records in the set of partial results for a resource.</param>
+        /// <param name="context">Serializer context</param>
+        /// <returns></returns>
+        public override Uri GenerateNextPageLink(Uri baseUri, int pageSize, Object instance, ODataSerializerContext context)
+        {
+            if (context == null || instance == null)
+            {
+                return null;
+            }
+
+            Func<object, string> skipTokenGenerator = null;
+            IList<OrderByNode> orderByNodes = null;
+            ExpandedReferenceSelectItem expandedItem = context.ExpandedSelectItem;
+            IEdmModel model = context.Model;
+
+            DefaultQuerySettings settings = context.QueryOptions.Context.DefaultQuerySettings;
+            if (settings.EnableSkipToken)
+            {
+                if (expandedItem != null)
+                {
+                    if (expandedItem.OrderByOption != null)
+                    {
+                        orderByNodes = OrderByNode.CreateCollection(expandedItem.OrderByOption);
+                    }
+
+                    skipTokenGenerator = (obj) =>
+                    {
+                        return GenerateSkipTokenValue(obj, model, orderByNodes);
+                    };
+                    
+                    return GetNextPageHelper.GetNextPageLink(baseUri, pageSize, instance, skipTokenGenerator);
+                }
+
+                if (context.QueryOptions != null && context.QueryOptions.OrderBy != null)
+                {
+                    orderByNodes = context.QueryOptions.OrderBy.OrderByNodes;
+                }
+
+                skipTokenGenerator = (obj) =>
+                {
+                    return GenerateSkipTokenValue(obj, model, orderByNodes);
+                };
+            }
+            return context.InternalRequest.GetNextPageLink(pageSize, instance, skipTokenGenerator);
```

Comment:
>return context.InternalRequest.GetNextPageLink(pageSize, instance, skipTokenGenerator); [](start = 12, length = 87)

That's executed out of "EnableSkipToken"?  Shall we return "null" for this or throw exception here?

Reply:
GetNextPageLink reverts to the prior behavior if skipTokenGenerator is null. EnableSkipToken determines if the skipTokenGenerator has to be set or not. 

---
In reply to: [282567877](https://github.com/OData/WebApi/pull/1724#discussion_r282567877) [](ancestors = 282567877)